analytic_diffuse/models.py

```diff
@@ -15,16 +17,54 @@ def checkinput(func):
     def wrapper(*args, **kwargs):
         # Check validity of input shapes
         phi = args[0]
+        # Check whether input was scalar.
+        scalar = np.isscalar(args[1])
+
         theta = args[1]
-        if not phi.shape == theta.shape:
+
+        if not scalar:
+            theta = np.array(theta)
+            if phi is not None:
+                phi = np.array(phi)
+
+        if scalar and theta > np.pi/2:
+            return 0
+
+        if phi is not None and not scalar and phi.shape != theta.shape:
             raise ValueError(
                 "phi and theta must have the same shape: {}, {}".format(
                     str(phi.shape), str(theta.shape)
                 )
             )
-        result = func(*args, **kwargs)
+        result = func(phi, theta, *args[2:], **kwargs)
+
         # Set pixels outside the horizon to zero.
-        result[theta > np.pi / 2] *= 0
+        if not scalar:
+            result[theta > np.pi / 2] = 0
+
+        # If a scalar was passed in, return a scalar.
+        return result
+    return wrapper
+
+
+def projected(func):
+    """Enforce input shapes and horizon cutoff."""
+    global _projected_funcs
+
+    _projected_funcs.append(func.__name__)
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        projected = kwargs.pop('projected', False)
```

Comment:
Will 'projected' ever be passed in as a keyword? I can't find any cases of it. If not, does this need to be a wrapper?

Reply:
Yes, this wasn't clear. Basically, the `projected` decorator added the `projected` keyword to its decorated functions, and enabled them to be returned in either projected or standard co-ordinates, but it wasn't documented anywhere. I've now added in documentation, and also expanded this to every function -- functions which are decorated with `projected` are _defined_ in projected space (but still by default return their models as unprojected). Models which are not decorated like this are _defined_ in standard co-ordinates (but can still optionally be returned as projected models).

The reason to do this is because the hankel transform uses the projected form, and for the models defined in projected space, this means redundantly multiplying then dividing by cos(za). This avoids that extra computation. 

I also added a `circsym` decorator to mark some functions as circularly symmetric (doesn't do anything, just adds them to a list for now).